### PR TITLE
Prevent accessing unicode scalar element with incorrect index.

### PR DESCRIPTION
**Pull request by sofiaromorales**
Bug/issue #, if applicable:  150760264

## Summary

This change prevents a potential DocC crash by making sure that it does not access an array element using an index that is outside of the bonds of the array.

`.index(after:` can crash in some cases if the passed index belongs to the last element of the collection. with this change we ensure that the passed index is not the last index.

## Testing

Run unit tests in Linux using Swift 5.9.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [X] Added tests
- [X] Ran the `./bin/test` script and it succeeded
- [(N/A)] Updated documentation if necessary

### CODE REVIEW

File: Sources/SwiftDocC/Utility/ValidatedURL.swift [hunk 1]
```diff
@@ -178,8 +178,13 @@ private extension StringProtocol {
     /// Returns a percent encoded version of the string or the original string if it is already percent encoded.
     func addingPercentEncodingIfNeeded(withAllowedCharacters allowedCharacters: CharacterSet) -> String? {
         var needsPercentEncoding: Bool {
-            for (index, character) in unicodeScalars.indexed() where !allowedCharacters.contains(character) {
+            let unicodeScalars = unicodeScalars.filter { !allowedCharacters.contains($0) }
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift:181]: ...or maybe we could just `dropLast(2)` on the sequence we're iterating over 🤔

File: Sources/SwiftDocC/Utility/ValidatedURL.swift [hunk 2]
```diff
@@ -178,8 +178,13 @@ private extension StringProtocol {
...
+            for (index, character) in unicodeScalars.indexed() {
                 if character == "%" {
+                    // There's not two characters after the "%". This "%" can't represent a percent encoded character.
+                    guard index != unicodeScalars.endIndex || index != unicodeScalars.index(before: index) else {
```
Comment (QuietMisdreavus) [inline Sources/SwiftDocC/Utility/ValidatedURL.swift:185]: `String.UnicodeScalarView.Index` is Comparable, so this can be simplified:

```suggestion
                    guard index < self.unicodeScalars.index(self.unicodeScalars.endIndex, offsetBy: -2, limitedBy: self.unicodeScalars.startIndex) else {
```

(The `self.` prefixes can be dropped if we go back to a `for/where` loop without a previous filter; i'm just including them here for completeness.)

File: Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift [hunk 3]
```diff
@@ -120,6 +120,13 @@ class ValidatedURLTests: XCTestCase {
             XCTAssertEqual(validatedWithHeading.components.path, expectedPath)
             XCTAssertEqual(validatedWithHeading.components.fragment, "Heading-Name")
         }
+        
+        for linkText in [
+            "url://com.example.test/new/docc=Whats%20New&version=DocC&Title=[Update]"
```
Comment (QuietMisdreavus) [inline Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift:125]: Do we have a test sample of a URL that isn't percent-encoded already? I would want to make sure that that is covered. (Plus a couple for short strings or other kinds of edge cases that this issue has uncovered.)

**Review by QuietMisdreavus — COMMENTED** [review]
Thanks for looking into this! I've got a couple comments to add on.

File: Sources/SwiftDocC/Utility/ValidatedURL.swift (hunk 2 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift:185]: I think this can also be expressed as  
```suggestion
                    guard unicodeScalars.distance(from: index, to: unicodeScalars.endIndex) <= 2 else {
```

Comment (sofiaromorales) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift:185]: or ` guard index < unicodeScalars.index(after: unicodeScalars.endIndex)`

Comment (sofiaromorales) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift:185]: or `guard index < unicodeScalars.index(before: unicodeScalars.endIndex)`

File: Sources/SwiftDocC/Utility/ValidatedURL.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (sofiaromorales) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift:181]: Yes. I added the `where` clause back. 🙏

**Comment by sofiaromorales** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Utility/ValidatedURL.swift [hunk 4]
```diff
@@ -203,7 +203,10 @@ private extension StringProtocol {
         return if needsPercentEncoding {
             addingPercentEncoding(withAllowedCharacters: allowedCharacters)
         } else {
-            String(self)
+            // Create a new string by mapping over the character helps
+            // preventing crashes when handling substrings that contains
+            // multibyte characters.
+            self.map { String($0) }.joined()
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Utility/ValidatedURL.swift:209]: Is this change necessary? 

If so, what are the performance implications of this? Reading this naively it seems like it could be horribly slow. But I'm not familiar with how Swift implements strings and sub-strings internally. We should run a benchmark to make sure this isn't a hot spot.

File: Sources/SwiftDocC/Utility/ValidatedURL.swift [hunk 5]
```diff
@@ -181,14 +181,14 @@ private extension StringProtocol {
             for (index, character) in unicodeScalars.indexed() where !allowedCharacters.contains(character) {
                 if character == "%" {
                     // % isn't allowed in a URL fragment but it is also the escape character for percent encoding.
-                    let firstFollowingIndex  = unicodeScalars.index(after: index)
-                    let secondFollowingIndex = unicodeScalars.index(after: firstFollowingIndex)
-                    
-                    guard secondFollowingIndex < unicodeScalars.endIndex else {
+                    guard unicodeScalars.distance(from: index, to: unicodeScalars.endIndex) <= 2 else {
```
Comment (QuietMisdreavus) [inline Sources/SwiftDocC/Utility/ValidatedURL.swift:184]: I feel like the sign on this is wrong? we want the distance to be *at least* 2 scalars from the end, whereas it just seems like this is checking for it to be *at most* 2 scalars.

```suggestion
                    guard unicodeScalars.distance(from: index, to: unicodeScalars.endIndex) >= 2 else {
```

File: Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift [hunk 6]
```diff
@@ -81,6 +81,14 @@ class ValidatedURLTests: XCTestCase {
     }
     
     func testQueryIsPartOfPathForAuthoredLinks() throws {
+        
+        func validate(linkText: String, expectedPath: String, expectedFragment: String? = nil) throws {
+            let validated = try XCTUnwrap(ValidatedURL(parsingAuthoredLink: linkText), "Failed to parse \(linkText.singleQuoted) as authored link")
+            XCTAssertNil(validated.components.queryItems, "Authored documentation links don't include query items")
+            XCTAssertEqual(validated.components.path, expectedPath)
+            XCTAssertEqual(validated.components.fragment, expectedFragment)
+        }
```
Comment (QuietMisdreavus) [inline Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift:90]: It's worth adding file/line arguments for assertion wrappers:

```suggestion
        func validate(linkText: String, expectedPath: String, expectedFragment: String? = nil, file: StaticString = #file, line: UInt = #line) throws {
            let validated = try XCTUnwrap(ValidatedURL(parsingAuthoredLink: linkText), "Failed to parse \(linkText.singleQuoted) as authored link", file: file, line: line)
            XCTAssertNil(validated.components.queryItems, "Authored documentation links don't include query items", file: file, line: line)
            XCTAssertEqual(validated.components.path, expectedPath, file: file, line: line)
            XCTAssertEqual(validated.components.fragment, expectedFragment, file: file, line: line)
        }
```

This way, assertion failures point to the caller of validate rather than the body.

**Review by QuietMisdreavus — CHANGES_REQUESTED** [review]
One request, one suggestion, and i'm also interested in @patshaughnessy's question.

File: Sources/SwiftDocC/Utility/ValidatedURL.swift (hunk 4 above)
Comment (sofiaromorales) [reply Sources/SwiftDocC/Utility/ValidatedURL.swift]: I changed it back.  Linux environments don't like constructing strings from substrings that contains multibyte characters and sometimes it results in crashes, but it's true that this has performance issues.

**Comment by sofiaromorales** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift (hunk 6 above)
Comment (sofiaromorales) [reply Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift]: Thanks for this!

File: Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift [hunk 7]
```diff
@@ -111,15 +113,24 @@ class ValidatedURLTests: XCTestCase {
...
-            XCTAssertEqual(validatedWithHeading.components.path, expectedPath)
-            XCTAssertEqual(validatedWithHeading.components.fragment, "Heading-Name")
+            try validate(linkText: linkText, expectedPath: expectedPath)
+            try validate(linkText: linkText + "#Heading-Name", expectedPath: expectedPath, expectedFragment: "Heading-Name")
         }
+        
+        // Test parameter with percent encoding
+        var linkText = "doc://com.example.test/docc=Whats%20New&version=DocC&Title=[Update]"
+        var expectedPath = "/docc=Whats%20New&version=DocC&Title=[Update]"
+        try validate(linkText: linkText, expectedPath: expectedPath)
```
Comment (anferbui) [inline Tests/SwiftDocCTests/Utility/ValidatedURLTests.swift:123]: Should we add a test for `%` being in the link but it being the final character or next to final character, to test the new logic?

**Review by anferbui — APPROVED** [review]
Approved with a comment about the tests

**Comment by sofiaromorales** [comment]
@swift-ci please test

**Review by d-ronnqvist — APPROVED** [review]
